File: testClients/JoshDougClient/UnitType.cpp
// -*-c++-*-

#include "UnitType.h"
#include "game.h"


UnitType::UnitType(_UnitType* pointer)
{
    ptr = (void*) pointer;
}

int UnitType::id()
{
  return ((_UnitType*)ptr)->id;
}

char* UnitType::name()
{
  return ((_UnitType*)ptr)->name;
}

int UnitType::type()
{
  return ((_UnitType*)ptr)->type;
}

int UnitType::cost()
{
  return ((_UnitType*)ptr)->cost;
}

int UnitType::attackPower()
{
  return ((_UnitType*)ptr)->attackPower;
}

int UnitType::digPower()
{
  return ((_UnitType*)ptr)->digPower;
}

int UnitType::fillPower()
{
  return ((_UnitType*)ptr)->fillPower;
}

int UnitType::maxHealth()
{
  return ((_UnitType*)ptr)->maxHealth;
}

int UnitType::maxMovement()
{
  return ((_UnitType*)ptr)->maxMovement;
}

int UnitType::offensePower()
{
  return ((_UnitType*)ptr)->offensePower;
}

int UnitType::defensePower()
{
  return ((_UnitType*)ptr)->defensePower;
}

int UnitType::range()
{
  return ((_UnitType*)ptr)->range;
}




std::ostream& operator<<(std::ostream& stream,UnitType ob)
{
  stream << "id: " << ((_UnitType*)ob.ptr)->id  <<'\n';
  stream << "name: " << ((_UnitType*)ob.ptr)->name  <<'\n';
  stream << "type: " << ((_UnitType*)ob.ptr)->type  <<'\n';
  stream << "cost: " << ((_UnitType*)ob.ptr)->cost  <<'\n';
  stream << "attackPower: " << ((_UnitType*)ob.ptr)->attackPower  <<'\n';
  stream << "digPower: " << ((_UnitType*)ob.ptr)->digPower  <<'\n';
  stream << "fillPower: " << ((_UnitType*)ob.ptr)->fillPower  <<'\n';
  stream << "maxHealth: " << ((_UnitType*)ob.ptr)->maxHealth  <<'\n';
  stream << "maxMovement: " << ((_UnitType*)ob.ptr)->maxMovement  <<'\n';
  stream << "offensePower: " << ((_UnitType*)ob.ptr)->offensePower  <<'\n';
  stream << "defensePower: " << ((_UnitType*)ob.ptr)->defensePower  <<'\n';
  stream << "range: " << ((_UnitType*)ob.ptr)->range  <<'\n';
  return stream;
}

File: shellAI/java/UnitType.java
import com.sun.jna.Pointer;

///Represents type of unit.
class UnitType
{
  Pointer ptr;
  int ID;
  int iteration;
  public UnitType(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.unitTypeGetId(ptr);
    iteration = BaseAI.iteration;
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.unitTypes.length; i++)
    {
      if(BaseAI.unitTypes[i].ID == ID)
      {
        ptr = BaseAI.unitTypes[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands


    //getters

  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.unitTypeGetId(ptr);
  }
  ///The name of this type of unit.
  public String getName()
  {
    validify();
    return Client.INSTANCE.unitTypeGetName(ptr);
  }
  ///The UnitType specific id representing this type of unit.
  public int getType()
  {
    validify();
    return Client.INSTANCE.unitTypeGetType(ptr);
  }
  ///The oxygen cost to spawn this unit type into the game.
  public int getCost()
  {
    validify();
    return Client.INSTANCE.unitTypeGetCost(ptr);
  }
  ///The power of the attack of this type of unit.
  public int getAttackPower()
  {
    validify();
    return Client.INSTANCE.unitTypeGetAttackPower(ptr);
  }
  ///The power of this unit types's digging ability.
  public int getDigPower()
  {
    validify();
    return Client.INSTANCE.unitTypeGetDigPower(ptr);
  }
  ///The power of this unit type's filling ability.
  public int getFillPower()
  {
    validify();
    return Client.INSTANCE.unitTypeGetFillPower(ptr);
  }
  ///The maximum amount of this health this unit can have
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.unitTypeGetMaxHealth(ptr);
  }
  ///The maximum number of moves this unit can move.
  public int getMaxMovement()
  {
    validify();
    return Client.INSTANCE.unitTypeGetMaxMovement(ptr);
  }
  ///The power of the unit type's offensive siege ability.
  public int getOffensePower()
  {
    validify();
    return Client.INSTANCE.unitTypeGetOffensePower(ptr);
  }
  ///The power of the unit type's defensive siege ability.
  public int getDefensePower()
  {
    validify();
    return Client.INSTANCE.unitTypeGetDefensePower(ptr);
  }
  ///The range of the unit type's attack.
  public int getRange()
  {
    validify();
    return Client.INSTANCE.unitTypeGetRange(ptr);
  }

}
